src/service/resources-impl.js

```diff
@@ -252,26 +254,53 @@ export class Resources {
    * @restricted
    */
   renderStarted() {
-    this.ampdoc.signals().signal('render-start');
+    this.ampdoc.signals().signal(CommonSignals.RENDER_START);
   }
 
   /**
-   * Returns a subset of resources which is identified as being in the current
-   * viewport.
+   * Returns a subset of resources which are (1) belong to the specified host
+   * window, and (2) positioned in the specified rect.
+   * @param {!Window} hostWin
+   * @param {!../layout-rect.LayoutRectDef} rect
    * @param {boolean=} opt_isInPrerender signifies if we are in prerender mode.
-   * @return {!Array<!Resource>}
+   * @return {!Promise<!Array<!Resource>>}
    */
-  getResourcesInViewport(opt_isInPrerender) {
+  getResourcesInRect(hostWin, rect, opt_isInPrerender) {
     opt_isInPrerender = opt_isInPrerender || false;
-    const viewportRect = this.viewport_.getRect();
-    return this.resources_.filter(r => {
-      if (r.hasOwner() || !r.isDisplayed() || !r.overlaps(viewportRect)) {
-        return false;
-      }
-      if (opt_isInPrerender && !r.prerenderAllowed()) {
-        return false;
-      }
-      return true;
+
+    // First, wait for the `ready-scan` signal. Waiting for each element
+    // individually is too expensive and `ready-scan` will cover most of
```

Comment:
> Waiting for each element individually is too expensive

@dvoytenko Why is that? do you mean the cost of creating many Promise instances?

I came across this code and wondered if we can get rid of `ready-scan` (which has a weird / unclear definition itself).

Reply:
This method was initially exclusively meant to cover performance measurement and thus tried to be as lightweight as possible. But I can't quite recollect how exactly this signal optimizes this. TBH, I'd think a bigger issue here is to wait for elements to be actually there to do this measurement at all. Otherwise it's very possible to make this call before any AMP elements have been parsed/added.

Reply:
I see. So you meant that we are actually relying on the side effect of this code (waiting for `documentReady_`, `ampInitialized_` & `first pass`) to wait for resources additions.

If that's the case, what do you think about switching to:
https://github.com/ampproject/amphtml/blob/185d4984314990b00e1af893b73d172261c2e1bc/src/service/resources-interface.js#L115

Reply:
That'd make sense to me. But not sure how much delta between them in practice.